File: inventory-service/src/infra/messagebroker/kafkaconfig.go
package messagebroker

import (
	"fmt"
	"os"

	"example.com/inventory-service/src/application/interfaces"
	"github.com/confluentinc/confluent-kafka-go/kafka"
)

type KafkaBroker struct {
	Producer *kafka.Producer
	Consumer *kafka.Consumer
}

func NewKafkaBroker() KafkaBroker {
	return KafkaBroker{
		Producer: ConfigKafkaProducer(),
		Consumer: ConfigKafkaConsumer(),
	}
}

func ConfigKafkaProducer() *kafka.Producer {
	p, err := kafka.NewProducer(&kafka.ConfigMap{
		"bootstrap.servers": "kafka:9092",
		"client.id":         "edms-group-1",
		"acks":              "all",
	})
	if err != nil {
		fmt.Printf("Failed to create producer: %s\n", err)
	}
	return p
}

func ConfigKafkaConsumer() *kafka.Consumer {
	c, err := kafka.NewConsumer(&kafka.ConfigMap{
		"bootstrap.servers": "kafka:9092",
		"group.id":          "edms-group-1",
	})
	if err != nil {
		fmt.Printf("Failed to configure c: %s\n", err)
	}
	return c
}

func (k KafkaBroker) SendToKafka(message string) {
	topic := "edms"
	delivery_chan := make(chan kafka.Event, 10000)
	err := k.Producer.Produce(&kafka.Message{
		TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
		Value:          []byte(message),
	},
		delivery_chan,
	)
	if err != nil {
		fmt.Printf("Failed to produce message: %s\n", err)
	}
	fmt.Println("Message sent to kafka")
}

func (k KafkaBroker) ConsumeFromKafka(messageBrokerService interfaces.IMessageBrokerConsumerService) {
	topics := []string{"edms"}
	err := k.Consumer.SubscribeTopics(topics, nil)
	run := err == nil

	for run == true {
		ev := k.Consumer.Poll(10000)
		switch e := ev.(type) {
		case *kafka.Message:
			messageBrokerService.ConsumeMessage(string(e.Value))
		case kafka.Error:
			fmt.Fprintf(os.Stderr, "%% Error: %v\n", e)
			run = false
		}
	}

	k.Consumer.Close()
}

File: payment-service/src/infra/message_broker/kafka_config.py
import os
from confluent_kafka import Consumer, Producer


class KafkaProducer:
    def __init__(self):
        self.producer = config_kafka_producer()

    def send_to_kafka(self, message):
        self.producer.produce("edms", message.encode("utf-8"))
        print("message sent to kafka: ", message)
        self.producer.flush()


class KafkaConsumer:
    def __init__(self):
        self.consumer = config_kafka_consumer()

    def consume_from_kafka(self):
        raw_message = self.consumer.poll(1.0)

        if raw_message is not None:
            message = raw_message.value().decode("utf-8")
            return message

        return

    def stop_consuming(self):
        self.consumer.close()


def config_kafka_producer():
    kafka_broker_address = os.environ.get("KAFKA_BROKER_ADDRESS")
    producer_config = {
        "bootstrap.servers": kafka_broker_address,
    }
    return Producer(producer_config)


def config_kafka_consumer():
    kafka_broker_address = os.environ.get("KAFKA_BROKER_ADDRESS")
    consumer_config = {
        "bootstrap.servers": kafka_broker_address,
        "group.id": "payment-service-consumer-group",
        "auto.offset.reset": "earliest",
    }

    consumer = Consumer(consumer_config)
    consumer.subscribe(["edms"])
    return consumer
